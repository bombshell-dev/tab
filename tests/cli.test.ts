import { exec } from "child_process";
import { describe, it, expect, test } from "vitest";

function runCommand(command: string): Promise<string> {
  return new Promise((resolve, reject) => {
    exec(command, (error, stdout, stderr) => {
      if (error) {
        reject(stderr);
      } else {
        resolve(stdout);
      }
    });
  });
}

const cliTools = ["citty"];

describe.each(cliTools)("cli completion tests for %s", (cliTool) => {
  const commandPrefix = `pnpm tsx demo.${cliTool}.ts complete --`;

  it("should complete cli options", async () => {
    const output = await runCommand(`${commandPrefix}`);
    expect(output).toMatchSnapshot();
  });

  describe("cli option completion tests", () => {
    const optionTests = [
      { partial: "--p", expected: "--port" },
    ];

    test.each(optionTests)(
      "should complete option for partial input '%s'",
      async ({ partial }) => {
        const command = `${commandPrefix} dev ${partial}`;
        const output = await runCommand(command);
        expect(output).toMatchSnapshot();
      }
    );
  });

  describe("cli option exclusion tests", () => {
    const alreadySpecifiedTests = [
      { specified: "--config", shouldNotContain: "--config" },
    ];

    test.each(alreadySpecifiedTests)(
      "should not suggest already specified option '%s'",
      async ({ specified }) => {
        const command = `${commandPrefix} ${specified} --`;
        const output = await runCommand(command);
        expect(output).toMatchSnapshot();
      }
    );
  });

  describe("cli option value handling", () => {
    it("should resolve port value correctly", async () => {
      const command = `${commandPrefix} dev --port=3`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });

    it("should handle conflicting options appropriately", async () => {
      const command = `${commandPrefix} --config vite.config.js --`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });

    it("should resolve config option values correctly", async () => {
      const command = `${commandPrefix} --config vite.config`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });

    it("should handle unknown options with no completions", async () => {
      const command = `${commandPrefix} --unknownoption`;
      const output = await runCommand(command);
      expect(output.trim()).toMatchSnapshot();
    });
  });

  describe("edge case completions for end with space", () => {
    //TOOD: remove this
    it("should suggest port values if user ends with space after `--port`", async () => {
      const command = `${commandPrefix} dev --port ""`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });

    it("should keep suggesting the --port option if user typed partial but didn't end with space", async () => {
      const command = `${commandPrefix} dev --po`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });

    it("should suggest port values if user typed `--port=` and hasn't typed a space or value yet", async () => {
      const command = `${commandPrefix} dev --port=`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });
  });

  // single positional command: `lint [file]`
  // vite ""
  // -> src/
  // -> ./

  // vite src/ ""
  // -> nothing
  // should not suggest anything

  // multiple postiionals command `lint [...files]`
  // vite ""
  // -> src/
  // -> ./

  // vite src/ ""
  // -> src/
  // -> ./

  describe("positional argument completions", () => {
    it("should complete single positional argument when ending with space (vite src/)", async () => {
      const command = `${commandPrefix} vite src/ ""`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });

    it("should complete multiple positional arguments when ending with space (vite src/ ./)", async () => {
      const command = `${commandPrefix} vite ""`;
      const output = await runCommand(command);
      expect(output).toMatchSnapshot();
    });
  });
});
