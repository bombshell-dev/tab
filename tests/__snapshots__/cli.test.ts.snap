// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli completion tests for cac > cli option completion tests > should complete option for partial input '{ partial: '--p', expected: '--port' }' 1`] = `
"--port	Specify port
:4
"
`;

exports[`cli completion tests for cac > cli option completion tests > should complete option for partial input '{ partial: '-H', expected: '-H' }' 1`] = `
":4
"
`;

exports[`cli completion tests for cac > cli option completion tests > should complete option for partial input '{ partial: '-p', expected: '-p' }' 1`] = `
":4
"
`;

exports[`cli completion tests for cac > cli option exclusion tests > should not suggest already specified option '{ specified: '--config', shouldNotContain: '--config' }' 1`] = `
":4
"
`;

exports[`cli completion tests for cac > cli option value handling > should handle unknown options with no completions 1`] = `":4"`;

exports[`cli completion tests for cac > cli option value handling > should not show duplicate options 1`] = `
"--config	Use specified config file
--mode	Set env mode
--logLevel	info | warn | error | silent
:4
"
`;

exports[`cli completion tests for cac > cli option value handling > should resolve config option values correctly 1`] = `
"vite.config.ts	Vite config file
vite.config.js	Vite config file
:4
"
`;

exports[`cli completion tests for cac > cli option value handling > should resolve port value correctly 1`] = `
"--port=3000	Development server port
:4
"
`;

exports[`cli completion tests for cac > edge case completions for end with space > should keep suggesting the --port option if user typed partial but didn't end with space 1`] = `
"--port	Specify port
:4
"
`;

exports[`cli completion tests for cac > edge case completions for end with space > should suggest port values if user ends with space after \`--port\` 1`] = `
"3000	Development server port
8080	Alternative port
:4
"
`;

exports[`cli completion tests for cac > edge case completions for end with space > should suggest port values if user typed \`--port=\` and hasn't typed a space or value yet 1`] = `
"--port=3000	Development server port
--port=8080	Alternative port
:4
"
`;

exports[`cli completion tests for cac > positional argument completions > should complete multiple positional arguments when ending with part of the value 1`] = `
"index.ts	Index file
:4
"
`;

exports[`cli completion tests for cac > positional argument completions > should complete multiple positional arguments when ending with space 1`] = `
"main.ts	Main file
index.ts	Index file
:4
"
`;

exports[`cli completion tests for cac > positional argument completions > should complete single positional argument when ending with space 1`] = `
"main.ts	Main file
index.ts	Index file
:4
"
`;

exports[`cli completion tests for cac > short flag handling > should handle global short flags 1`] = `
":4
"
`;

exports[`cli completion tests for cac > short flag handling > should handle short flag value completion 1`] = `
":4
"
`;

exports[`cli completion tests for cac > short flag handling > should handle short flag with equals sign 1`] = `
":4
"
`;

exports[`cli completion tests for cac > short flag handling > should not show duplicate options when short flag is used 1`] = `
"--config	Use specified config file
--mode	Set env mode
--logLevel	info | warn | error | silent
:4
"
`;

exports[`cli completion tests for cac > should complete cli options 1`] = `
"dev	Start dev server
serve	Start the server
lint	Lint project
:4
"
`;

exports[`cli completion tests for citty > cli option completion tests > should complete option for partial input '{ partial: '--p', expected: '--port' }' 1`] = `
"--port	Specify port
:4
"
`;

exports[`cli completion tests for citty > cli option completion tests > should complete option for partial input '{ partial: '-H', expected: '-H' }' 1`] = `
"-H	Specify hostname
:4
"
`;

exports[`cli completion tests for citty > cli option completion tests > should complete option for partial input '{ partial: '-p', expected: '-p' }' 1`] = `
"-p	Specify port
:4
"
`;

exports[`cli completion tests for citty > cli option exclusion tests > should not suggest already specified option '{ specified: '--config', shouldNotContain: '--config' }' 1`] = `
":4
"
`;

exports[`cli completion tests for citty > cli option value handling > should handle unknown options with no completions 1`] = `":4"`;

exports[`cli completion tests for citty > cli option value handling > should not show duplicate options 1`] = `
"--config	Use specified config file
--mode	Set env mode
--logLevel	info | warn | error | silent
:4
"
`;

exports[`cli completion tests for citty > cli option value handling > should resolve config option values correctly 1`] = `
"vite.config.ts	Vite config file
vite.config.js	Vite config file
:4
"
`;

exports[`cli completion tests for citty > cli option value handling > should resolve port value correctly 1`] = `
"--port=3000	Development server port
:4
"
`;

exports[`cli completion tests for citty > edge case completions for end with space > should keep suggesting the --port option if user typed partial but didn't end with space 1`] = `
"--port	Specify port
:4
"
`;

exports[`cli completion tests for citty > edge case completions for end with space > should suggest port values if user ends with space after \`--port\` 1`] = `
"3000	Development server port
8080	Alternative port
:4
"
`;

exports[`cli completion tests for citty > edge case completions for end with space > should suggest port values if user typed \`--port=\` and hasn't typed a space or value yet 1`] = `
"--port=3000	Development server port
--port=8080	Alternative port
:4
"
`;

exports[`cli completion tests for citty > short flag handling > should handle global short flags 1`] = `
":4
"
`;

exports[`cli completion tests for citty > short flag handling > should handle short flag value completion 1`] = `
"-p	Specify port
:4
"
`;

exports[`cli completion tests for citty > short flag handling > should handle short flag with equals sign 1`] = `
":4
"
`;

exports[`cli completion tests for citty > short flag handling > should not show duplicate options when short flag is used 1`] = `
"--config	Use specified config file
--mode	Set env mode
--logLevel	info | warn | error | silent
:4
"
`;

exports[`cli completion tests for citty > should complete cli options 1`] = `
"dev	Start dev server
build	Build project
lint	Lint project
:4
"
`;

exports[`cli completion tests for commander > cli option value handling > should handle unknown options with no completions 1`] = `":4"`;

exports[`cli completion tests for commander > short flag handling > should handle global short flags 1`] = `
"-c	specify config file
:4
"
`;

exports[`cli completion tests for commander > should complete cli options 1`] = `
"serve	Start the server
build	Build the project
deploy	Deploy the application
lint	Lint source files
:4
"
`;
